File: scripts/utils/antigravity_diagnostics.py
"""Antigravity CLI failure diagnostics shared by research-digest callers."""
from __future__ import annotations

from pathlib import Path
from typing import Any

GLOBAL_FAILURE_REASONS = frozenset({"quota_exhausted", "auth_required", "startup_failed"})


def diag_reason(runner) -> str | None:
    """读 runner 最近一次失败诊断的 reason（无 last_diagnostics 属性的注入型 runner 返 None）。

    与 build_diagnostics（写侧）配对的读侧入口。board_break/pk、string_yang/mainline 等
    调用方按 reason 分流（如 timeout 不重试）；此处收拢 getattr/isinstance 守卫，避免各
    caller 内联重复该读法（与 runner 构造侧的 4 处同构一并属 invoke_llm_cli 收敛 defer）。
    """
    diag = getattr(runner, "last_diagnostics", None)
    return diag.get("reason") if isinstance(diag, dict) else None


def build_diagnostics(
    *,
    stdout: str,
    stderr: str,
    log_file: str | Path | None = None,
    reason: str | None = None,
    returncode: int | None = None,
) -> dict[str, Any]:
    log_text = read_text_if_exists(log_file)
    combined = "\n".join(part for part in (stdout or "", stderr or "", log_text or "") if part)
    detected_reason = classify_failure(combined, default=reason or "antigravity_failed")
    out: dict[str, Any] = {
        "reason": detected_reason,
        "message": diagnostic_message(combined, reason=detected_reason) or detected_reason,
        "stdout_empty": not bool((stdout or "").strip()),
        "stderr_empty": not bool((stderr or "").strip()),
    }
    if log_file:
        out["log_file"] = str(log_file)
    if returncode is not None:
        out["returncode"] = returncode
    return out


def classify_failure(text: str, *, default: str) -> str:
    lower = (text or "").lower()
    if (
        "resource_exhausted" in lower
        or "code 429" in lower
        or "quota reached" in lower
        or "quota exceeded" in lower
        or "individual quota" in lower
    ):
        return "quota_exhausted"
    if "not a valid artifact path" in lower or "invalid tool call error" in lower:
        return "agent_tool_error"
    if (
        not _auth_recovered(lower)
        and (
            "not logged in" in lower
            or "not authenticated" in lower
            or "unauthenticated" in lower
            or "opening authentication page" in lower
            or "authorization code" in lower
        )
    ):
        return "auth_required"
    if default == "startup_failed":
        return "startup_failed"
    if "timeout" in lower:
        return "timeout"
    return default


def diagnostic_message(text: str, *, reason: str | None = None) -> str:
    if not text:
        return ""
    preferred_by_reason = {
        "quota_exhausted": ("RESOURCE_EXHAUSTED", "code 429", "quota reached", "quota exceeded", "individual quota"),
        "auth_required": ("not logged in", "not authenticated", "authentication", "authorization code"),
        "agent_tool_error": ("not a valid artifact path", "invalid tool call error", "model output error"),
        "timeout": ("timeout", "timed out"),
    }
    preferred = preferred_by_reason.get(reason or "", (
        "RESOURCE_EXHAUSTED",
        "code 429",
        "quota reached",
        "quota exceeded",
        "individual quota",
        "not a valid artifact path",
        "invalid tool call error",
        "not logged in",
        "not authenticated",
        "authentication",
        "timeout",
    ))
    lines = [line.strip() for line in text.splitlines() if line.strip()]
    for line in lines:
        lower = line.lower()
        if any(term.lower() in lower for term in preferred):
            return line[:500]
    return lines[0][:500] if lines else ""


def _auth_recovered(lower_text: str) -> bool:
    return (
        "auth succeeded" in lower_text
        or "applyauthresult:" in lower_text
        or "experiments refreshed after login" in lower_text
    )


def read_text_if_exists(path: str | Path | None) -> str:
    if not path:
        return ""
    try:
        p = Path(path)
        if not p.exists():
            return ""
        return p.read_text(encoding="utf-8", errors="replace")[:32000]
    except OSError:
        return ""


def is_global_failure(reason: str | None) -> bool:
    return str(reason or "") in GLOBAL_FAILURE_REASONS

File: scripts/workflows/antigravity_diagnostics.mjs
import fs from "node:fs";

export const GLOBAL_FAILURE_REASONS = new Set(["quota_exhausted", "auth_required", "startup_failed"]);

export function diagnoseAntigravityFailure({
  stdout = "",
  stderr = "",
  logFile = "",
  reason = "antigravity_failed",
  returncode = null,
} = {}) {
  const logText = readTextIfExists(logFile);
  const combined = [stdout, stderr, logText].filter(Boolean).join("\n");
  const detected = classifyAntigravityFailure(combined, reason);
  const diagnostics = {
    reason: detected,
    message: diagnosticMessage(combined) || detected,
    stdout_empty: !String(stdout || "").trim(),
    stderr_empty: !String(stderr || "").trim(),
  };
  if (logFile) diagnostics.log_file = logFile;
  if (returncode !== null && returncode !== undefined) diagnostics.returncode = returncode;
  return diagnostics;
}

export function classifyAntigravityFailure(text, fallback = "antigravity_failed") {
  const lower = String(text || "").toLowerCase();
  if (lower.includes("resource_exhausted") || lower.includes("code 429") || lower.includes("quota")) {
    return "quota_exhausted";
  }
  if (
    lower.includes("not logged in")
    || lower.includes("not authenticated")
    || lower.includes("unauthenticated")
    || lower.includes("opening authentication page")
    || lower.includes("authorization code")
  ) {
    return "auth_required";
  }
  if (fallback === "startup_failed") return "startup_failed";
  if (lower.includes("timeout")) return "timeout";
  return fallback || "antigravity_failed";
}

export function diagnosticMessage(text) {
  const lines = String(text || "").split(/\r?\n/).map((line) => line.trim()).filter(Boolean);
  const preferred = ["resource_exhausted", "code 429", "quota", "not logged in", "not authenticated", "authentication", "timeout"];
  for (const line of lines) {
    const lower = line.toLowerCase();
    if (preferred.some((term) => lower.includes(term))) {
      return line.slice(0, 500);
    }
  }
  return lines.length ? lines[0].slice(0, 500) : "";
}

export function readTextIfExists(file) {
  if (!file) return "";
  try {
    if (!fs.existsSync(file)) return "";
    return fs.readFileSync(file, "utf-8").slice(0, 4000);
  } catch (_) {
    return "";
  }
}

export function sanitizeDiagnostics(value) {
  if (!value || typeof value !== "object" || Array.isArray(value)) return {};
  const out = {};
  for (const key of ["reason", "message", "log_file", "returncode", "stdout_empty", "stderr_empty"]) {
    if (value[key] !== undefined && value[key] !== null && value[key] !== "") out[key] = value[key];
  }
  return out;
}

export function isGlobalLlmFailure(reason) {
  return GLOBAL_FAILURE_REASONS.has(String(reason || ""));
}
